cellarium/ml/transforms/transforms.py

```diff
@@ -1,108 +1,261 @@
 from torch import nn
 
+from cellarium.ml.utilities.testing import (
+    assert_arrays_equal,
+    assert_columns_and_array_lengths_equal,
+    assert_nonnegative,
+    assert_positive,
+)
 
-class NormalizeTotal(nn.Module):
+
+class DivideByScale(nn.Module):
     """
-    Normalize total gene counts per cell to target count.
+    Divide gene counts by a scale.
+
+    .. math::
+
+        y_{ng} = \\frac{x_{ng}}{\\mathrm{scale}_g + \\mathrm{eps}}
 
     Args:
-        target_count:
-            Target gene epxression count.
+        scale_g:
+            A scale for each gene.
+        feature_schema:
+            The variable names schema for the input data validation.
         eps:
             A value added to the denominator for numerical stability.
     """
 
-    def __init__(
-        self,
-        target_count: int = 10_000,
-        eps: float = 1e-6,
-    ):
+    def __init__(self, scale_g: torch.Tensor, feature_schema: ArrayLike, eps: float = 1e-6) -> None:
         super().__init__()
-        self.target_count = target_count
+        self.scale_g: torch.Tensor
+        self.register_buffer("scale_g", scale_g)
+        self.feature_schema = np.array(feature_schema)
+        assert_nonnegative("eps", eps)
         self.eps = eps
 
+    def forward(
+        self,
+        x_ng: torch.Tensor,
+        feature_g: np.ndarray | None = None,
+    ) -> torch.Tensor:
+        """
+        Args:
+            x_ng:
+                Gene counts.
+            feature_g:
+                The list of the variable names in the input data. If ``None``, no validation is performed.
+
+        Returns:
+            Gene counts divided by scale.
+        """
+        if feature_g is not None:
+            assert_columns_and_array_lengths_equal("x_ng", x_ng, "feature_g", feature_g)
+            assert_arrays_equal("feature_g", feature_g, "feature_schema", self.feature_schema)
+
+        return x_ng / (self.scale_g + self.eps)
+
+    def __repr__(self) -> str:
+        return (
+            f"{self.__class__.__name__}(scale_g={self.scale_g}, feature_schema={self.feature_schema}, eps={self.eps})"
+        )
+
+
+class Filter(nn.Module):
+    """
+    Filter gene counts by a list of features.
+
+    .. math::
+
+        \\mathrm{mask}_g = \\mathrm{feature}_g \\in \\mathrm{filter\\_list}
+
+        y_{ng} = x_{ng}[:, \\mathrm{mask}_g]
+
+    Args:
+        filter_list: A list of features to filter by.
+    """
+
+    def __init__(self, filter_list: ArrayLike) -> None:
+        super().__init__()
+        self.filter_list = np.array(filter_list)
+        if len(self.filter_list) == 0:
+            raise ValueError(f"`filter_list` must not be empty. Got {self.filter_list}")
+
+    @cache
+    def filter(self, feature_g: tuple) -> np.ndarray[Any, np.dtype[np.bool_]]:
+        """
+        Args:
+            feature_g: The list of the variable names in the input data.
+
+        Returns:
+            A boolean mask of the features to filter by.
```

Comment:
tuple(ndarray.tolist()) is a bottleneck. If feature_g (ndarray) is hashable, let's not transform it here and let filter do it.